### ISSUE TITLE
task error handling

### ISSUE THREAD

**Issue by jdx** [jdx/mise#3525]
there should be some kind of error handling to execute something when a task fails, perhaps to cancel the error somehow. My use-case for this is I have a bunch of tasks that run and leave the project in an inconsistent state when they fail because it's all running in parallel. There is a specific task that—if failed—I'd like it to reset the contents of a specific file.

Noodling on this:

```toml
[tasks.render]
run = "node build-something.js > output"
on_failure = "git checkout output"
```

I don't personally need to cancel the error—though I feel like with this sort of thing that's probably going to be behavior someone wants—especially if I call this "error handling".

Maybe we could make that an optional setting:

```toml
[tasks.render]
run = "node build-something.js > output"
on_failure = {run = "git checkout output", cancel_error = true}
```

what do other task managers do?

**Comment by jdx** [comment]
perhaps a better way to name this would be "cleanup" and not worry about catching errors

**Comment by acesyde** [comment]
Maybe add the ability to only trigger the failure case on a specific exit code ?

**Comment by jdx** [comment]
I was going to pass the status code as an env var

**Comment by reteps** [comment]
The status code should be available in $? in my opinion for use in cleanup.

**Comment by jdx** [comment]
setting that before a script runs is complicated and something that would have to be unique depending on which shell is being used. Imagine a handler being written in python, for example.

**Comment by reteps** [comment]
Ah, you are correct. TIL: https://stackoverflow.com/a/26627346

**Comment by egnor** [comment]
You wouldn't just do something like this?

```
run = "node build-something.js > output || (git checkout output && false)"
```

...at some level a task runner isn't a full scripting language?

In general I'm a little suspicious of "cleanup" steps since if you control-C at the top level or something else happens you won't get your cleanup, so it's kind of a brittle thing to count on -- I would tend to encourage

```
run = [
  "node build-something.js > output.tmp",
  "mv output.tmp output",
]
```

but of course there are lots of examples where that's not so easy.

**Comment by reitzig** [comment]
> what do other task managers do?

Taskfile: "nice" syntax is open issue [go-task/task#141](https://github.com/go-task/task/issues/141), though it's _possible_ with `defer` (cf. [go-task/task#141.1826783326](https://github.com/go-task/task/issues/141#issuecomment-1826783326))

**Comment by jdx** [comment]
Covered by `depends_post`, which now runs even when the parent task fails (#8274), so it can be used for cleanup. A failure-only hook can be a new issue if it's still needed.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-opus-5-5; version: unavailable.*
